src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+
+        #region - AddGraphTypes -
+        /// <summary>
+        /// Scans the calling assembly for classes that implement <see cref="IGraphType"/> and registers
+        /// them as transients within the dependency injection framework. A transient lifetime ensures
+        /// they are only instianted once each time the schema is built. If the schema is a scoped schema,
+        /// the graph types will effectively be scoped graph types. If the schema is a singleton schema,
+        /// the graph types will effectively be singleton graph types.
+        /// <br/><br/>
+        /// Also registers <see cref="EnumerationGraphType{TEnum}"/>, <see cref="ConnectionType{TNodeType}"/>,
+        /// <see cref="ConnectionType{TNodeType, TEdgeType}"/>, <see cref="EdgeType{TNodeType}"/>,
+        /// <see cref="InputObjectGraphType{TSourceType}"/>, <see cref="AutoRegisteringInputObjectGraphType{TSourceType}"/>, and
+        /// <see cref="AutoRegisteringObjectGraphType{TSourceType}"/> as generic types.
+        /// </summary>
+        public static IGraphQLBuilder AddGraphTypes(this IGraphQLBuilder builder)
+            => builder.AddGraphTypes(Assembly.GetCallingAssembly());
+
+        /// <summary>
+        /// Scans the supplied assembly for classes that implement <see cref="IGraphType"/> and registers
+        /// them as transients within the dependency injection framework. A transient lifetime ensures
+        /// they are only instianted once each time the schema is built. If the schema is a scoped schema,
+        /// the graph types will effectively be scoped graph types. If the schema is a singleton schema,
+        /// the graph types will effectively be singleton graph types.
+        /// <br/><br/>
+        /// Also registers <see cref="EnumerationGraphType{TEnum}"/>, <see cref="ConnectionType{TNodeType}"/>,
+        /// <see cref="ConnectionType{TNodeType, TEdgeType}"/>, <see cref="EdgeType{TNodeType}"/>,
+        /// <see cref="InputObjectGraphType{TSourceType}"/>, <see cref="AutoRegisteringInputObjectGraphType{TSourceType}"/>, and
+        /// <see cref="AutoRegisteringObjectGraphType{TSourceType}"/> as generic types.
+        /// </summary>
+        public static IGraphQLBuilder AddGraphTypes(this IGraphQLBuilder builder, Assembly assembly)
+        {
+            // Graph types are always created with the transient lifetime, since they are only instianted once
+            // each time the schema is built. If the schema is a scoped schema, the graph types will effectively
+            // be scoped graph types. If the schema is a singleton schema, the graph types will effectively be
+            // singleton graph types. This is REQUIRED behavior and must not be changed.
+
+            if (assembly == null)
+                throw new ArgumentNullException(nameof(assembly));
+
+            foreach (var type in assembly.GetTypes()
+                .Where(x => x.IsClass && !x.IsAbstract && typeof(IGraphType).IsAssignableFrom(x)))
```

Comment:
It may be useful to provider some attribute to exclude types from registration (like `DoNotMapClrTypeAttribute`).

Reply:
Since it is DI registration, usually excessive registrations do not hurt anything.  I have used code like this for a long time and never found any need to exclude a type.  Note also that `TryRegister` is used for all types.

Reply:
> usually

Usually, not always.

Reply:
#2798